File: generated_code/py/program_1_3.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # Map to store the value and its corresponding index: {value: index}
        # This allows us to achieve an optimal O(n) time complexity by trading O(n) space.
        seen = {}
        
        for index, num in enumerate(nums):
            complement = target - num
            
            # If the complement exists in our map, we found the pair
            if complement in seen:
                return [seen[complement], index]
            
            # Otherwise, store the current number and its index for future lookups
            seen[num] = index
            
        # The problem guarantees exactly one solution, so we should never reach here.
        return []

File: generated_code/cpp/program_1_3.cpp
#include <vector>
#include <unordered_map>

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        // Use a hash map to store the value and its corresponding index.
        // This allows us to look up the complement in O(1) average time.
        std::unordered_map<int, int> num_to_index;
        
        for (int i = 0; i < nums.size(); ++i) {
            int complement = target - nums[i];
            
            // Check if the complement already exists in the map
            auto it = num_to_index.find(complement);
            if (it != num_to_index.end()) {
                // If found, return the index of the complement and the current index
                return {it->second, i};
            }
            
            // Otherwise, store the current number and its index in the map
            num_to_index[nums[i]] = i;
        }
        
        // The problem guarantees exactly one solution, so this point is unreachable.
        return {};
    }
};